README.md

```diff
@@ -345,6 +345,24 @@ included by default unless you choose which one to exclude with the
 then you have to specify each backend that you want individually. You can get
 the full list of these flags with the `--help` flag.
 
+#### --auto-config-uri
+This is a special URI which if set, will try and pull a config from that
+location on startup. It will use the cookie file stored at
+`--auto-config-cookie-path` if specified. If successful, it will check if the
+config is different from what is currently stored. If so then it will validate
+if it is a valid json config. If so it will replace (overwrite!) the current
+config and then run with that!
```

Comment:
I could get behind caching the remote config, but overwriting the user's own config would be a surprising result with unfortunate consequences (see related comments to `main.go`). Exclamations in the README will *explain* why the unexpected happened, but by then it's too late (nobody reads the README until they have to).

Reply:
nit: @mattsb42-aws pointed me to https://sembr.org/ and while it can be a bit cumbersome at first to try to use line breaks that way it can make things a lot more readable, and changes to text easier to review.

Reply:
None of the auto-config stuff happens automatically (and as you say surprisingly) unless the user explicitly turns it on somehow. Only then will it do its magic.

Reply:
I mean that if I pass an `--auto-config-uri` I don't know that it's going to overwrite my own config unless I read the README before doing that. We should do the least surprising thing the user can expect based on what they know without reading.

Reply:
Having a tool load a remote configuration, overwrite the local configuration, and execute that new configuration all at once would be very surprising to me. Instead of trying to do everything all at once, I suggest following the Unix philosophy of "Do One Thing Well." This has the dual benefits of simplifying your user experience and simplifying your implementation logic.

How I would break this out in this particular case is:

- execute : Reads the local config file and executes based on the contents of that config.
- config sync : Loads a remote config, parses it to verify that it is valid, and overwrites the local config file.

bonus points for:

- config diff / config sync dryrun : Loads a remote config, parses it to verify that it is valid, loads the local config, parses _it_ to verify that it is valid, and display the diff between the two.

Reply:
Based on my reading of the code, I think @amznpurple's goal was to make it so that a user can pass `--auto-config-uri` once and only once, thereafter only passing `--config-path` and ceding control to whoever manages the config at the URI (or URIs that URI points to). That's what makes it an "auto" config URI rather than just a `config-uri`.

That can make sense if the config at the URI doesn't set `config-path`, because `--config-path <path-to-config>` can be shorter than `--config-uri https://example.com/yesiscan/config.json`.

Things also get weird if the URI points to another URI, but the user continues to pass `--auto-config-uri` because command-line options will override config values. A URI might point to another URI in order to upgrade users to a later version of the config, but the user who doesn't know to stop using the flag will never see the new config.

Things also git weird if that URI points to another URI, or if the user doesn't pass `--config-path` but the config at the URI sets `config-path` to anything other than the default.

Things get dangerous if a bad actor obtains control, even temporarily, of the URI and sets `config-path` to a file that is both writable and important, or sets `auto-config-uri` to a URI they have full control of, but only if the user doesn't pass `--config-path` and / or `--auto-config-uri`.

I'm not sure I see a way out of this with the current (recursive) design. There are definitely dangers inherent in letting a `config.json` specify a `config-path` and then also using recursion to load configs. They're mitigating by the user passing `--config-path`, but then why not make that the only way to set the path (along with `--profile`). It's also scary that a config at a URI can set or unset `quiet`.

I do like the idea of passing `--auto-config-uri` once, and having subsequent executions of `yesiscan` check that URI for updates. But similar to Matt's suggestion, it would be better if that was a command and not a flag, so that setting the URI is something you do explicitly once (or once per profile) and not something you could also do while running the tool.

I'm mixed on letting the config at a URI change the URI. It seems like HTTP redirection is probably a better way to handle that (possibly including writing the new location if we reach a permanent redirect).

I don't like recursion as a way to implement this, especially since that implementation detail can leak as program behavior (such as writing files) that's visible to the user.

This is a case where designing the customer visible behavior we actually want, and then implementing to that design, would be better. Do we want these URIs to be able to chain to each other? What should the end state be in that case? Do we want it to be possible for that URI to overwrite config values the user has set? Should it be possible for `yesiscan` to overwrite multiple files? Or files that aren't in its own `~/.config/` path?

Reply:
> Having a tool load a remote configuration, overwrite the local configuration, and execute that new configuration all at once would be very surprising to me.

The feature is intended for use my someone who doesn't know anything about local config. In fact, we can even bake in that flag into the actual binary. So they just run the thing once, and it works. So the fact that it's overwriting their config is fine.

> How I would break this out in this particular case is:
>    execute : Reads the local config file and executes based on the contents of that config.
>    config sync : Loads a remote config, parses it to verify that it is valid, and overwrites the local config file.

These are exactly the steps that are occurring in the code posted.

> bonus points for:
> display the diff between the two.

I agree that would be cool. Since this is meant to mostly be a hidden detail from the user, I don't expect to need it, but we can consider it in the future.

Reply:
As one of your users, I can only speak for myself, but I am also surprised by the idea that I should supply a `--config-path` and then also not "know anything about local config".

> this is meant to mostly be a hidden detail from the user

If that's a goal, can we make it so that the path is also unknown to the user? For example maybe the `--auto-config-uri` option could require the `--profile` option, and then the config at the URI could be required to include its own URI as the `auto-config-uri` and to not include a `config-path`.

Can you say more about `why` that's a goal?